### ISSUE TITLE
Discussion: pattern in foreach 

### ISSUE THREAD

**Issue by alrz** [dotnet/csharplang#1437]
Ability to use a pattern in `foreach`
```cs
foreach (case StatementSyntax statement in nodes) {}
foreach (case {Result: var result} in tasks) {} 
```
->
```cs
foreach (SyntaxNode node in nodes) {
  if (!(node is StatementSyntax statement)) continue;
}

foreach (Task<T> task in tasks) {
  if (!(task is {Result: var result})) continue;
}
```

**Comment by Starwer** [comment]
@alrz: I'm not sure this makes it really more clear, but I agree on the sweetness of a pattern matching in a foreach-loop.
I would feel this is more natural:

```C#
foreach (var statement is StatementSyntax in nodes) {}
foreach (var task is {Result: var result} in tasks) {} 
```

**Comment by jnm2** [comment]
This is common for me. With new pattern matching features, I expect it to become even more common.

**Comment by Richiban** [comment]
Something similar is likely to "fall out" for free if we get both [`do` query termination](https://github.com/dotnet/csharplang/issues/101) and pattern matching in Linq expressions:

```csharp
from node in nodes
where node is StatementSyntax statement
do {
    // Do something with `statement`
}
```

```csharp
from task in tasks
where task is {Result: var result} 
do {
    // Use result here 
} 
```

**Comment by alrz** [comment]
@Richiban For `is`/`out` in query expressions, see https://github.com/dotnet/csharplang/issues/159.

**Comment by VanKrock** [comment]
May be use linq?
``` cs
foreach (var statement in nodes.OfType<StatementSyntax>()) 
{
}

foreach (var result in tasks.Select(t => t.Result)) 
{
} 
```

**Comment by Starwer** [comment]
@VanKrock  : Linq is the "nice" way to do it nowadays, but it is more expensive in CPU and memory than what would the examples given by @alrz . So, although you examples are decently readable, the proposal of @alrz  would still be an improvement.

**Comment by alrz** [comment]
Note that `out var` and pattern variables do not work in query expressions like that. as of 7.3, they are allowed but the scope doesn't leak out to the next operator so you need to manually create tuples and propagate values (see the linked issue for the discussion).

This syntax, on the other hand, would provide the same conciseness that we have with deconstructions in foreach,

```cs
foreach (var (x, y) in listOfTuples)
```
You can see the deconstruction as a "var pattern" that always succeed with a tuple. now if you have a pattern that isn't "complete" (may fail to macth) you add `case` to skip over those.

```cs
foreach (case var (x, y) in list OfNullableTuples)
```
Of course, you can imagine any other patterns to be used here.

**Comment by VanKrock** [comment]
@Starwer How much more expensive?

**Comment by VanKrock** [comment]
@Starwer 
``` cs
public static IEnumerable<TResult> Select<T, TResult>(this IEnumerable<T> @this, Func<T, TResult> selector)
{
    foreach (var item in @this)
        yield return selector.Invoke(item);
}
```

yield return works without temp collection  and create only `IEnumerator<TResult>` as i know.
``` cs
foreach (var result in tasks.Select(t => t.Result)) 
{
    ...
} 

-->

foreach (var task in tasks) 
{
    var result = task.Result;
    ...
} 
```

**Comment by Starwer** [comment]
@VanKrock : Thanks. Then I'm happily surprised about the way it is done when using LINQ. I then just cannot guess how expensive it is, as it depend how well the compiler optimizer would unfold the nested sub-routine calls. Sorry for my too assertive comment.

**Comment by Richiban** [comment]
@alrz Did the team drop the idea of having a `let... else...` statement? I seem to remember them floating the idea at one point but I think they may have let it go.

This idea if very closely related, i.e. "Here's a control statement where I introduce some variables via a pattern; if the pattern fails then skip the statement." I like it, but it would be weird to only support it in `foreach` statements.

Can we complement it with a 

```csharp
let case (0, y) = SomeValue();
else return; // Nothing to do here
```

**Comment by alrz** [comment]
@Richiban Yeah I guess that idea was dropped at some point..

I thought about something like that but you could do the same thing with "inverted if" feature (#882),
```cs
if !(SomeValue() is <pattern>) return;
```
That would make the alternatives less verbose:
```cs
foreach (var node in nodes) {
  if !(node is <pattern>) continue;
}
```
But since we already have deconstruction in foreach, I think it'd make sense to extend it to fallible patterns.

**Comment by Starwer** [comment]
Maybe could we also consider a Query-like syntax. This would be more verbose, but more versatile (not only bound to pattern-matching), and still quite readable:

```C#
foreach (from <decl> in <IEnumerable> where <expr>)
{
    <block>
} 
```
equivalent to:
```C#
foreach (var <decl> in <IEnumerable>)
{
    if (!(<expr>)) continue;

    <block>
} 
```

**Example:**
```C#
foreach (from node in nodes where node is StatementSyntax statement) 
{
    // 'node' and 'statement' are defined
}

foreach (from task in tasks where task is {Result: var result}) 
{
    // 'task' and 'result' are defined
}
```

To keep things consistent, we could enable also more advanced queries, by having the `in` of the foreach implicit when there is `from` in place:
```C#
foreach (var result from task in tasks where task.Result != null select task.Result) 
{
    // only 'result' is define. Would be nice if 'task' could be defined as well...
}
```

**Comment by Starwer** [comment]
I was wondering why my previous deliberately shocking proposal got so little reaction. It is probably because this has been discussed a million time and came out that the query-foreach syntax must be #101.
So, please forgive my "not new idea". Yet, I still like the balance conciseness/flexibility that such a construct would have been:
```C#
foreach (from <decl> in <IEnumerable> where <expr>) {}
```

**Comment by Starwer** [comment]
@bondsbw point taken. Let me rephrase that: I like the way it look. I feel the query style, without the parentheses, looks like an alien in the C-family world... That feels... inconsistent.
To keep things together, Shouldn't we remove the `foreach` totally and write the loops this way in the future ?:
```C#
from <decl> in <IEnumerable> do {}
```

**Comment by bondsbw** [comment]
@Starwer Actually I might use that in the future.

Removing `foreach` would break everybody's code, so of course that won't happen.  In hindsight maybe LINQ should have used `foreach` instead of `from` so they could eventually be combined, but that ship sailed many years ago.

**Comment by NN---** [comment]
There is some POC to rewrite linq into more optimized code: https://github.com/antiufo/roslyn-linq-rewrite

**Comment by NN---** [comment]
Pattern matching in foreach is very handy.
For instance Nemerle language has it: https://github.com/rsdn/nemerle/wiki/Foreach-features

**Comment by Thaina** [comment]
It not as bad as create new collection but it still create and use delegate

Pattern based is interesting that it would be faster. But I prefer linq syntax because pattern is making a mess on the left side

If linq would just be inlined it would be the best

**Comment by alrz** [comment]
Closing in favor of https://github.com/dotnet/csharplang/issues/1522
